iFixit/vigilo
# Feat: Add Bundler and tighten Docker Image

## Description

This has some major changes to the build logic and dev environment. The core logic remains the same.

### Changes:
- Restructured the files in the repo to be more granular with the content
- Added `tsup` and `tsc-alias` to compile and bundle the code for specific uses and targets
- Added `path` aliases in the `tsconfig.json`
- Added new helper scripts to get Devs setup quickly and run docker

### CR

The main goal of these changes were to allow us to have individual folders for our Typescript code -- i.e. config, core, and scripts. Then when we needed to execute the code, it would be properly compiled without nested folders in the `dist` folder. To accomplish that, we needed a bundler that could parse all the Typescript code and compile it for Node as standalone JS files.

I tried out different bundlers: `rspack`, `swc`, and `webpack`. They all had their own issues and I couldn't get them to work. This was mainly due to a Node issue with ES modules and CJS modules not being able to be mixed but also not being able to be separated when working on a Node app:
https://github.com/evanw/esbuild/issues/1921#issuecomment-1784828307.

At the end of the day I only got `esbuild` to work and in the desired way to address both #11 and #14. That said, `esbuild` still had its own issues which led me to discover the tool [`tsup`](https://tsup.egoist.dev). `tsup` is a wrapper around `esbuild` and fixes some of the underlying issues of working with the `esbuild` API including shimming modules -- https://tsup.egoist.dev/#inject-cjs-and-esm-shims. 

Aside from that, we also need the ability to not include the config files -- i.e `lh-config.js` and `urls.json` -- in the bundled `dist/index.js` file in order to manually configure these options for the Docker container. If they were included then we would have to rebuild the Docker image every time we wanted to change the configs and would need to build and tag multiple images for variations in the configs. d9d6942775db809cd28249de74ca6411abadf8d0 provides further details and explanation around this requirement.

### QA

- [ ] The `setup` script works
	1. Run `pnpm run setup`
	2. Confirm the prior templated files (`.env`, `lh-config.js`, and `urls.json`) have been copied to the appropriate locations.
- [ ] The `run-docker` script works
	1. Follow the `README` setup instructions
	2. Build the docker image `docker build -t vigilo .`
	3. Run `pnpm run start:docker`
- [ ] Confirm the `.env` file is not copied over to the docker image
	1. Follow the `README` setup instructions
	2. Build the docker image `docker build -t vigilo .`
	3. Run `docker run --rm -it vigilo /bin/bash`
	4. `ls` to ensure the `.env` file was not copied
- [ ] `start` script still works - Reference the instructions from https://github.com/iFixit/vigilo/pull/1
- [ ] `updateDatadogMetricsMetadata` script still works - Reference the instructions from https://github.com/iFixit/vigilo/pull/9
- [ ] `createDashboards` script still works - Reference the instructions from https://github.com/iFixit/vigilo/pull/8

Closes: #11
Closes: #14 

cr_req 2
